### ISSUE TITLE
Storing images in memory instead of on disk.

### ISSUE THREAD

**Issue by pranavbaburaj** [pranavbaburaj/img#11]
### Discussed in https://github.com/pranavbaburaj/img/discussions/9

<div type='discussions-op-text'>

<sup>Originally posted by **AnonymouX47** October  5, 2021</sup>
This applies in three ways:

1. A `DrawImage` instance holds a reference to a PIL image instead of a filepath.
   - This way, the file doesn't have to be read from disk every time the image is to be drawn.
   - This will easily allow support for cases where an image is being processed and only currently exists in memory e.g PIL image objects and more.
   - Increases the speed of repeated image drawing.
   - Of course, this also comes at the cost of memory but i think the trade-off is worth it.
   - Makes the library usable in an environment where file writes are either not possible or not wanted (maybe on some servers or cloud-based systems).
   - **The other points below are hinged upon this one.**
2. Images downloaded from URLs are stored as PIL image objects instead of files.
   - This is dependent upon and shares the benefits/liabilities of (1.).
3. Frames of GIFs are stored as PIL images instead of files.
   - This is dependent upon and shares the benefits/liabilities of (1.).
   - I think this is good because the frames are only used once, so why take them from memory to disk, then back to memory.
   - Drastically increases speed.
   - Reduces amount of disk read/write.

Let me know what you think... I've already checked (and tested) and I'm sure 2 and 3 are possible.</div>

**Comment by pranavbaburaj** [comment]
The reason I store the images(after downloading) in the disk is because the `draw_image` function uses a filename containing the object and `from_url` uses the `draw_image` function to draw the image. I think we can change it to something in which the user passes in an image object? What do you think?

```py
img = DrawImage(Image.open(filename))
img.draw_image()
```

**Comment by AnonymouX47** [comment]
> The reason I store the images(after downloading) in the disk is because the `draw_image` function uses a filename containing the object and `from_url` uses the `draw_image` function to draw the image.

Exactly, i understand... that's why i said storing the downloaded images as `Image` objects instead of on-disk files will depend on **No. 1** i.e changing `Drawimage` objects to store references to image objects and of course, the change will also have to reflect in the methods.

> I think we can change it to something in which the user passes in an image object? What do you think?
> 
> ```py
> img = DrawImage(Image.open(filename))
> img.draw_image()
> ```

Yes, it's actually one of the reasons i suggested this idea:

> - This will easily allow support for cases where an image is being processed and only currently exists in memory e.g PIL image objects and more.

but instead of making it the only form of input, it'll simply be another option which can go two ways:
- the image can be passed to the constructor just like you suggest above and type-checking will be used to determine if the argument is a filepath or a PIL image.
- a `from_pil_image()` method can be added.
- **I think i prefer the first option.**

So `DrawImage` will have three forms of input:
- Local filepath
- Image URL
- PIL images

**Comment by AnonymouX47** [comment]
I would suggest that this will be better in a separate major release (e.g 0.2.0).

The fixes and changes in in #8 other other previous ones can be added in an incremental release (e.g 0.1.1).

**Comment by pranavbaburaj** [comment]
Hmm. In any case, we will have to make changes with the `draw_image` function I guess. 
Maybe, we can move the `draw_image` to a private function with a PIL image as the parameter. All the three methods(`draw_image`, `from_url`, `from_pil_image`) will use this function to draw the image. 

```py
def draw_image(self, filename):
    # validation....
    self.__draw_image(Image.open(filename)))
```

But, in the case of `from_url`, we can use `Image.frombytes` to convert the raw bytes to an image.
Please let me know your thoughts :)

**Comment by AnonymouX47** [comment]
This was how I imagined it:

The attribute `__filename` changes to `__image` and now references a PIL Image already converted to RGB.
For instance, within `__init__()`, we'll have something like:

```python
self.__image  = image.convert("RGB")
```

* * *

I think I like the idea of a separate method that all it does is to print an image to console.
It'll come in handy for GIF frames, so there'll be no need to create a `DrawImage` instance per frame i.e

```python
def __display_gif(self):
    # ...
    self.__draw_image(self.__image.convert("RGB").resize(self.size))
    # ...
```

* * *

For downloaded images, `Image.frombytes()` can only be used if the image size and colour mode is known before hand, which I don't think is possible.
Instead, what I propose (which I've tested) is:

```python
def from_url(...):
    # process URL, download image
    return __class__(Image.open(io.BytesIO(response.content)), size)

```

**Comment by pranavbaburaj** [comment]
That makes sense. Thanks. let's implement it by the next release :)

**Comment by pranavbaburaj** [comment]
The issue has been hopefully fixed in [`997944c`](https://github.com/pranavbaburaj/img/commit/997944cc9be313e437826d239701fe76c3e395db). Please don't hesitate to reopen the issue if you are having issues with the new changes. Hoping to hear from you. Closing the issue for now :)

**Comment by AnonymouX47** [comment]
Concerning this issue... while I was testing before implementation, I realized it's a **really** bad idea...

I loaded a 9000x6357 image of ~1.9MB and it used up about 460MB of RAM.

Thinking about it, it's a really bad idea because when an image is loaded, it's decompressed and each pixel the same amount of space and every other one, then that size multiplied by the height and width :sad:... catastrophy!

**Comment by AnonymouX47** [comment]
So far the changes i've made on my side are:
- Added the `__draw_image()` method that converts the pixel data to a string containing the entire code and text to be printed.
  - At the same time changing the method of printing to processing the entire image into control-codes and text before writing to console. This yielded positive performance differences.
- Changed `__display_gif()` to pass the frames to the new `__draw_image()` directly instead of saving frames on-disk and then creating new `DrawImage` objects and loading each frame back again on subsequent iterations.
  - The performance difference is quite significant (more noticeable as _size_ gets larger).

**Comment by pranavbaburaj** [comment]
Ah nice. I would like to hear if you suggest moving back to storing images on the disk? .I don't think using a lot of RAM is a nice solution. What do you think 🤔 ?

**Comment by AnonymouX47** [comment]
> Ah nice. I would like to hear if you suggest moving back to storing images on the disk? .I don't think using a lot of RAM is a nice solution. What do you think 🤔 ?

Yeah, I do... i just edited my previous reply.

**Comment by pranavbaburaj** [comment]
> So far the changes i've made on my side are:
> 
> * Added the `__draw_image()` method that converts the pixel data to a string containing the entire code and text to be printed.
>   
>   * At the same time changing the method of printing to processing the entire image into control-codes and text before writing to console. This yielded positive performance differences.
> * Changed `__display_gif()` to pass the frames to the new `__draw_image()` directly instead of saving frames on-disk and then creating new `DrawImage` objects and loading each frame back again on subsequent iterations.
>   
>   * The performance difference is quite significant (more noticeable as _size_ gets larger).
> 
> EDIT: Now, i think it's best to just store file-paths for file-path and URL input (i.e download images on-disk) and then store the image objects for PIL image input. Then in `draw_image()`, the type of object stored can be used to differentiate between them i.e
> 
> ```python
> def __init__(self, source: Union[str, Image.Image], ...):
>     # Validity check
>     self.__source = source
>     # ...
> 
> def draw_image(self):
>     if isinstance(self.__source, Image.Image):
>         print(self.__draw_image(self.__source))
>     else:
>         # The same process as before.
> ```
> 
> Please let me know what you think...

That looks nice. I think we can use this. :D

**Comment by pranavbaburaj** [comment]
> > The issue has been hopefully fixed in [`997944c`](https://github.com/pranavbaburaj/img/commit/997944cc9be313e437826d239701fe76c3e395db). Please don't hesitate to reopen the issue if you are having issues with the new changes. Hoping to hear from you. Closing the issue for now :)
> 
> So, I guess we should revert this commit, right? 🤔

Yeah. I have merged the pull request :). Also, Sorry for the late reply.

**Comment by pranavbaburaj** [comment]
I think we have decided not to store images in memory. So, I think we can close the issue. Don't hesitate to reopen the issue If you have any suggestions on this. Thank you :D
